File: services/bff-water-planning/src/core/rid_calendar.py
"""Administrative irrigation calendar and source-driven crop activity."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date, timedelta
from enum import Enum

CONTRACT_SET_SHA256 = "987006c64402e79b9cb7af29358283f4cb368203cbd46935490a7ff453115539"

ANCHOR_MONTH = 11
ANCHOR_DAY = 1
WEEKS_PER_IRRIGATION_YEAR = 53
MIN_IRRIGATION_YEAR_CE = 1901
MAX_IRRIGATION_YEAR_CE = 2401
BE_OFFSET = 543


def _check_civil_date(value: date, name: str) -> None:
    if type(value) is not date:
        raise TypeError(f"{name} must be a datetime.date, not {type(value).__name__}")


@dataclass(frozen=True)
class IrrigationYear:
    ce: int
    be: int

    def __post_init__(self) -> None:
        if (
            isinstance(self.ce, bool)
            or not isinstance(self.ce, int)
            or isinstance(self.be, bool)
            or not isinstance(self.be, int)
        ):
            raise TypeError("irrigation year CE and BE values must be integers")
        if not MIN_IRRIGATION_YEAR_CE <= self.ce <= MAX_IRRIGATION_YEAR_CE:
            raise ValueError(
                f"irrigation year CE {self.ce} outside supported range "
                f"{MIN_IRRIGATION_YEAR_CE}..{MAX_IRRIGATION_YEAR_CE}"
            )
        if self.be != self.ce + BE_OFFSET:
            raise ValueError("irrigation year BE must equal CE + 543")

    @classmethod
    def from_ce(cls, value: int) -> IrrigationYear:
        if isinstance(value, bool) or not isinstance(value, int):
            raise TypeError("irrigation year CE value must be an integer")
        return cls(ce=value, be=value + BE_OFFSET)

    @classmethod
    def from_be(cls, value: int) -> IrrigationYear:
        if isinstance(value, bool) or not isinstance(value, int):
            raise TypeError("irrigation year BE value must be an integer")
        return cls(ce=value - BE_OFFSET, be=value)


@dataclass(frozen=True)
class DateSpan:
    start: date
    end: date

    def __post_init__(self) -> None:
        _check_civil_date(self.start, "start")
        _check_civil_date(self.end, "end")
        if self.end < self.start:
            raise ValueError("date span end must not precede start")


@dataclass(frozen=True)
class IrrigationWeek:
    irrigation_year: IrrigationYear
    irrigation_week: int

    def __post_init__(self) -> None:
        if type(self.irrigation_year) is not IrrigationYear:
            raise TypeError("irrigation_year must be an IrrigationYear")
        if isinstance(self.irrigation_week, bool) or not isinstance(
            self.irrigation_week,
            int,
        ):
            raise TypeError("irrigation_week must be an integer")
        if not 1 <= self.irrigation_week <= WEEKS_PER_IRRIGATION_YEAR:
            raise ValueError(
                "irrigation_week must be within " f"1..{WEEKS_PER_IRRIGATION_YEAR}"
            )

    @property
    def key(self) -> str:
        return f"{self.irrigation_year.ce:04d}-R{self.irrigation_week:02d}"


class CropActivityState(str, Enum):
    NOT_PLANTED = "not_planted"
    ACTIVE = "active"
    HARVESTED = "harvested"


@dataclass(frozen=True)
class CropActivity:
    state: CropActivityState
    crop_week: int | None

    def __post_init__(self) -> None:
        if type(self.state) is not CropActivityState:
            raise TypeError("state must be a CropActivityState")
        if self.state is CropActivityState.ACTIVE:
            if (
                isinstance(self.crop_week, bool)
                or not isinstance(self.crop_week, int)
                or self.crop_week < 1
            ):
                raise ValueError("active crop activity requires a positive crop week")
        elif self.crop_week is not None:
            raise ValueError("inactive crop activity must not carry a crop week")


def _irrigation_year_span(identity: IrrigationYear) -> DateSpan:
    return DateSpan(
        start=date(identity.ce - 1, ANCHOR_MONTH, ANCHOR_DAY),
        end=date(identity.ce, 10, 31),
    )


def irrigation_year(day: date) -> IrrigationYear:
    """Return the ending-year CE/BE identity containing the civil date."""
    _check_civil_date(day, "day")
    ending_year = (
        day.year + 1 if (day.month, day.day) >= (ANCHOR_MONTH, ANCHOR_DAY) else day.year
    )
    return IrrigationYear.from_ce(ending_year)


def irrigation_week(day: date) -> IrrigationWeek:
    """Return the administrative irrigation week containing the civil date."""
    identity = irrigation_year(day)
    offset = (day - _irrigation_year_span(identity).start).days
    return IrrigationWeek(
        irrigation_year=identity,
        irrigation_week=offset // 7 + 1,
    )


def irrigation_week_span(week: IrrigationWeek) -> DateSpan:
    """Return the inclusive span of an irrigation week."""
    if type(week) is not IrrigationWeek:
        raise TypeError("week must be an IrrigationWeek")
    year_span = _irrigation_year_span(week.irrigation_year)
    start = year_span.start + timedelta(days=7 * (week.irrigation_week - 1))
    return DateSpan(start=start, end=min(start + timedelta(days=6), year_span.end))


def crop_activity(
    planting_date: date,
    expected_harvest_date: date,
    on: date,
) -> CropActivity:
    """Return crop state and week for a source-provided active window."""
    _check_civil_date(planting_date, "planting_date")
    _check_civil_date(expected_harvest_date, "expected_harvest_date")
    _check_civil_date(on, "on")
    if expected_harvest_date < planting_date:
        raise ValueError("expected harvest date must not precede planting date")
    if on < planting_date:
        return CropActivity(CropActivityState.NOT_PLANTED, None)
    if on > expected_harvest_date:
        return CropActivity(CropActivityState.HARVESTED, None)
    return CropActivity(
        CropActivityState.ACTIVE,
        (on - planting_date).days // 7 + 1,
    )

File: services/ros/src/utils/rid-calendar.js
'use strict';

/** Administrative irrigation calendar and source-driven crop activity. */

const CONTRACT_SET_SHA256 =
  '987006c64402e79b9cb7af29358283f4cb368203cbd46935490a7ff453115539';

const ANCHOR_MONTH = 11;
const ANCHOR_DAY = 1;
const WEEKS_PER_IRRIGATION_YEAR = 53;
const MIN_IRRIGATION_YEAR_CE = 1901;
const MAX_IRRIGATION_YEAR_CE = 2401;
const BE_OFFSET = 543;
const MS_PER_DAY = 86400000;
const ISO_DATE = /^(\d{4})-(\d{2})-(\d{2})$/;

const pad = (value, width) => String(value).padStart(width, '0');

function toCivilDate(value, name) {
  if (value instanceof Date) {
    if (Number.isNaN(value.getTime())) {
      throw new RangeError(`${name} is an invalid Date`);
    }
    if (value.getTime() % MS_PER_DAY !== 0) {
      throw new RangeError(
        `${name} must be exactly UTC midnight; an instant is not a civil date`,
      );
    }
    return value.toISOString().slice(0, 10);
  }
  if (typeof value !== 'string') {
    throw new TypeError(
      `${name} must be a YYYY-MM-DD string or a UTC-midnight Date`,
    );
  }
  const match = ISO_DATE.exec(value);
  if (!match) {
    throw new RangeError(
      `${name} must be formatted YYYY-MM-DD, got ${JSON.stringify(value)}`,
    );
  }
  const [, year, month, day] = match.map(Number);
  const roundTrip = new Date(Date.UTC(year, month - 1, day));
  if (roundTrip.toISOString().slice(0, 10) !== value) {
    throw new RangeError(`${name} is not a real calendar date: ${value}`);
  }
  return value;
}

const toEpochDay = (civilDate) =>
  Date.UTC(
    ...civilDate
      .split('-')
      .map(Number)
      .map((part, index) => (index === 1 ? part - 1 : part)),
  ) / MS_PER_DAY;

const fromEpochDay = (epochDay) =>
  new Date(epochDay * MS_PER_DAY).toISOString().slice(0, 10);

function irrigationYearFromCe(ce) {
  if (!Number.isInteger(ce)) {
    throw new TypeError('irrigation year CE value must be an integer');
  }
  if (ce < MIN_IRRIGATION_YEAR_CE || ce > MAX_IRRIGATION_YEAR_CE) {
    throw new RangeError(
      `irrigation year CE ${ce} outside supported range ` +
        `${MIN_IRRIGATION_YEAR_CE}..${MAX_IRRIGATION_YEAR_CE}`,
    );
  }
  return Object.freeze({ ce, be: ce + BE_OFFSET });
}

function irrigationYearFromBe(be) {
  if (!Number.isInteger(be)) {
    throw new TypeError('irrigation year BE value must be an integer');
  }
  return irrigationYearFromCe(be - BE_OFFSET);
}

const IrrigationYear = Object.freeze({
  fromBe: irrigationYearFromBe,
  fromCe: irrigationYearFromCe,
});

function validateIrrigationYear(value) {
  if (
    !value ||
    !Number.isInteger(value.ce) ||
    !Number.isInteger(value.be) ||
    value.be !== value.ce + BE_OFFSET
  ) {
    throw new TypeError('irrigationYear must carry matching CE and BE values');
  }
  IrrigationYear.fromCe(value.ce);
}

function irrigationYearSpan(identity) {
  validateIrrigationYear(identity);
  return {
    start: `${pad(identity.ce - 1, 4)}-${pad(ANCHOR_MONTH, 2)}-${pad(
      ANCHOR_DAY,
      2,
    )}`,
    end: `${pad(identity.ce, 4)}-10-31`,
  };
}

function irrigationYear(day) {
  const civilDate = toCivilDate(day, 'day');
  const [year, month, dayOfMonth] = civilDate.split('-').map(Number);
  const endingYear =
    month > ANCHOR_MONTH || (month === ANCHOR_MONTH && dayOfMonth >= ANCHOR_DAY)
      ? year + 1
      : year;
  return IrrigationYear.fromCe(endingYear);
}

function irrigationWeek(day) {
  const civilDate = toCivilDate(day, 'day');
  const year = irrigationYear(civilDate);
  const offset =
    toEpochDay(civilDate) - toEpochDay(irrigationYearSpan(year).start);
  const week = Math.floor(offset / 7) + 1;
  return Object.freeze({
    irrigationYear: year,
    irrigationWeek: week,
    key: `${pad(year.ce, 4)}-R${pad(week, 2)}`,
  });
}

function irrigationWeekSpan(week) {
  if (
    !week ||
    !Number.isInteger(week.irrigationWeek) ||
    week.irrigationWeek < 1 ||
    week.irrigationWeek > WEEKS_PER_IRRIGATION_YEAR
  ) {
    throw new RangeError(
      `irrigationWeek must be within 1..${WEEKS_PER_IRRIGATION_YEAR}`,
    );
  }
  const yearSpan = irrigationYearSpan(week.irrigationYear);
  const start = fromEpochDay(
    toEpochDay(yearSpan.start) + 7 * (week.irrigationWeek - 1),
  );
  const end = fromEpochDay(
    Math.min(toEpochDay(start) + 6, toEpochDay(yearSpan.end)),
  );
  return {
    start,
    end,
    lengthDays: toEpochDay(end) - toEpochDay(start) + 1,
  };
}

function cropActivity(plantingDate, expectedHarvestDate, on) {
  const planting = toCivilDate(plantingDate, 'plantingDate');
  const harvest = toCivilDate(expectedHarvestDate, 'expectedHarvestDate');
  const observed = toCivilDate(on, 'on');
  const plantingDay = toEpochDay(planting);
  const harvestDay = toEpochDay(harvest);
  const observedDay = toEpochDay(observed);

  if (harvestDay < plantingDay) {
    throw new RangeError(
      'expected harvest date must not precede planting date',
    );
  }
  if (observedDay < plantingDay) {
    return { state: 'not_planted', cropWeek: null };
  }
  if (observedDay > harvestDay) {
    return { state: 'harvested', cropWeek: null };
  }
  return {
    state: 'active',
    cropWeek: Math.floor((observedDay - plantingDay) / 7) + 1,
  };
}

module.exports = {
  CONTRACT_SET_SHA256,
  IrrigationYear,
  cropActivity,
  irrigationWeek,
  irrigationWeekSpan,
  irrigationYear,
};
